File: examples/python/textual/notes_showcase/src/notes_showcase/viewmodels/notifications_vm.py
"""NotificationsVM — bounded mirror of an INotificationHub's pending list.

Subscribes to ``INotificationHub.pending`` and surfaces the most recent
notifications as a bounded :class:`vmx.ObservableList` of
:class:`vmx.notifications.NotificationVM`. Cap defaults to 5 (drops the
oldest entry when full); auto-dismiss is handled by each inner
``NotificationVM``'s lifespan timer.
"""

from __future__ import annotations

import dataclasses
from datetime import timedelta

from reactivex.abc import DisposableBase, SchedulerBase
from reactivex.scheduler import TimeoutScheduler

from vmx import (
    ComponentVM,
    MessageHub,
    MessageHubProto,
    ObservableList,
    PropertyChangedMessage,
    RxDispatcher,
)
from vmx.messages.protocols import Message
from vmx.notifications import INotificationHub, Notification, NotificationVM
from vmx.services.dispatcher import Dispatcher

_DEFAULT_CAP = 5


class NotificationsVM(ComponentVM):
    """Bounded mirror of an INotificationHub's pending notifications."""

    DEFAULT_CAP = _DEFAULT_CAP

    def __init__(
        self,
        *,
        name: str,
        hint: str,
        hub: MessageHubProto[Message],
        dispatcher: Dispatcher,
        notification_hub: INotificationHub,
        scheduler: SchedulerBase | None = None,
        lifespan: timedelta | None = None,
        cap: int = _DEFAULT_CAP,
    ) -> None:
        super().__init__(name=name, hint=hint, hub=hub, dispatcher=dispatcher)
        self._notification_hub = notification_hub
        self._scheduler = scheduler or TimeoutScheduler()
        self._lifespan = lifespan
        self._cap = cap
        self._visible: ObservableList[NotificationVM] = ObservableList()
        self._map: dict[Notification, NotificationVM] = {}
        self._pending_sub: DisposableBase | None = None

    # ── Public surface ─────────────────────────────────────────────────────
    @property
    def hub(self) -> MessageHubProto[Message]:
        return self._hub

    @property
    def visible(self) -> ObservableList[NotificationVM]:
        return self._visible

    @property
    def cap(self) -> int:
        return self._cap

    # ── Lifecycle ──────────────────────────────────────────────────────────
    def _on_construct(self) -> None:
        super()._on_construct()
        self._pending_sub = self._notification_hub.pending.subscribe(
            on_next=self._sync_from_pending
        )

    def _on_destruct(self) -> None:
        if self._pending_sub is not None:
            self._pending_sub.dispose()
            self._pending_sub = None
        self._clear_visible()
        super()._on_destruct()

    def _on_dispose(self) -> None:
        if self._pending_sub is not None:
            self._pending_sub.dispose()
            self._pending_sub = None
        self._clear_visible()
        super()._on_dispose()

    def _clear_visible(self) -> None:
        for vm in list(self._visible):
            vm.dispose()
        with self._visible.batch_update():
            while self._visible.count > 0:
                self._visible.remove_at(0)
        self._map.clear()

    def _sync_from_pending(self, pending: list[Notification]) -> None:
        # Add VMs for notifications not yet rendered.
        for n in pending:
            if n in self._map:
                continue
            vm = NotificationVM(
                notification=n,
                hub=self._notification_hub,
                scheduler=self._scheduler,
                lifespan=self._lifespan,
            )
            self._map[n] = vm
            self._visible.append(vm)
            # Drop oldest while over cap.
            while self._visible.count > self._cap:
                oldest = self._visible[0]
                self._visible.remove_at(0)
                key = next((k for k, v in self._map.items() if v is oldest), None)
                if key is not None:
                    self._map.pop(key, None)
                oldest.dispose()
        # Remove VMs whose notifications are no longer pending.
        still = set(map(id, pending))
        to_remove = [n for n in self._map if id(n) not in still]
        for n in to_remove:
            vm = self._map.pop(n)
            self._visible.remove(vm)
            vm.dispose()
        self._hub.send(PropertyChangedMessage.create(self, self._name, "visible"))
        self._raise_property_changed("visible")

    # ── Builder entry-point ────────────────────────────────────────────────
    @staticmethod
    def builder() -> NotificationsVMBuilder:  # type: ignore[override]
        # Narrows ComponentVM.builder() to the showcase NotificationsVMBuilder.
        return NotificationsVMBuilder()


@dataclasses.dataclass(frozen=True, slots=True)
class NotificationsVMBuilder:
    """Immutable fluent builder for :class:`NotificationsVM`."""

    _name: str | None = None
    _hint: str = ""
    _hub: MessageHubProto[Message] | None = None
    _dispatcher: Dispatcher | None = None
    _notification_hub: INotificationHub | None = None
    _scheduler: SchedulerBase | None = None
    _lifespan: timedelta | None = None
    _cap: int = _DEFAULT_CAP

    def name(self, value: str) -> NotificationsVMBuilder:
        return dataclasses.replace(self, _name=value)

    def hint(self, value: str) -> NotificationsVMBuilder:
        return dataclasses.replace(self, _hint=value)

    def services(
        self, hub: MessageHubProto[Message], dispatcher: Dispatcher
    ) -> NotificationsVMBuilder:
        return dataclasses.replace(self, _hub=hub, _dispatcher=dispatcher)

    def notification_hub(self, value: INotificationHub) -> NotificationsVMBuilder:
        return dataclasses.replace(self, _notification_hub=value)

    def scheduler(self, value: SchedulerBase) -> NotificationsVMBuilder:
        return dataclasses.replace(self, _scheduler=value)

    def lifespan(self, value: timedelta) -> NotificationsVMBuilder:
        return dataclasses.replace(self, _lifespan=value)

    def cap(self, value: int) -> NotificationsVMBuilder:
        return dataclasses.replace(self, _cap=value)

    def build(self) -> NotificationsVM:
        if self._name is None:
            raise ValueError("name is required")
        if self._notification_hub is None:
            raise ValueError("notification_hub is required")
        hub = self._hub if self._hub is not None else MessageHub[Message]()
        dispatcher = (
            self._dispatcher
            if self._dispatcher is not None
            else RxDispatcher.immediate()
        )
        return NotificationsVM(
            name=self._name,
            hint=self._hint,
            hub=hub,
            dispatcher=dispatcher,
            notification_hub=self._notification_hub,
            scheduler=self._scheduler,
            lifespan=self._lifespan,
            cap=self._cap,
        )

File: examples/swift/notes-showcase/Sources/NotesShowcaseCore/ViewModels/NotificationsVM.swift
//
// NotificationsVM — bounded, keyed notification collection.
//
// Ports examples/csharp/avalonia/NotesShowcase/ViewModels/NotificationsVM.cs.
// See task-6-brief.md.
//
// Subscribes to `notificationHub.pending` on construct (foreground-marshalled
// via the injected dispatcher). Maintains a bounded `[NotificationVM]` list
// (`visible`, capped at `cap`, default 5). Deduplicates by Notification
// identity and removes resolved notifications when the hub no longer lists them.
//
// Cross-module subclassing enabled by ADR-0066: `hub`, `dispatcher`, and
// `_notifyPropertyChanged` are `public` on `ComponentVMBase`.
//
import Foundation
import Combine
import VMx

/// Bounded view-model list for in-flight notifications.
///
/// Subscribes to the notification hub on construct and removes the
/// subscription on destruct / dispose. Each element is a `NotificationVM`
/// whose lifespan timer is driven by the injected `VirtualTimeScheduler`.
public final class NotificationsVM: ComponentVMBase {

    /// Default maximum concurrently rendered notifications.
    public static let defaultCap = 5

    // ── Private state ──────────────────────────────────────────────────────

    private let _notificationHub: NotificationHubProtocol
    private let _scheduler: VirtualTimeScheduler
    private let _lifespan: TimeInterval?
    private let _cap: Int

    private var _visible: [NotificationVM] = []
    /// Keyed by `ObjectIdentifier(notification)` for O(1) dedup.
    private var _map: [ObjectIdentifier: NotificationVM] = [:]
    private var _pendingCancellable: AnyCancellable?

    // ── Public surface ─────────────────────────────────────────────────────

    /// Bounded list of currently-rendered notifications.
    ///
    /// Updated synchronously on the foreground thread; observers may re-read
    /// this after receiving a `"visible"` `PropertyChangedMessage` on the hub.
    public var visible: [NotificationVM] { _visible }

    /// Maximum number of concurrently rendered notifications.
    public var cap: Int { _cap }

    // ── Init ───────────────────────────────────────────────────────────────

    private init(
        name: String,
        hint: String,
        hub: MessageHubProtocol,
        dispatcher: Dispatcher,
        notificationHub: NotificationHubProtocol,
        scheduler: VirtualTimeScheduler,
        lifespan: TimeInterval?,
        cap: Int
    ) {
        _notificationHub = notificationHub
        _scheduler = scheduler
        _lifespan = lifespan
        _cap = cap
        super.init(name: name, hint: hint, hub: hub, dispatcher: dispatcher)
    }

    // ── Lifecycle overrides ────────────────────────────────────────────────

    /// Subscribes to `notificationHub.pending` on construct.
    ///
    /// The pending stream (backed by `CurrentValueSubject`) immediately emits
    /// the current snapshot on subscription — `syncFromPending` runs once with
    /// the current state before the first external post.
    ///
    /// Posts arrive from background async continuations and timer callbacks;
    /// `syncFromPending` mutates `_visible` which is observed by the view.
    /// Marshal every sync to the foreground dispatcher to keep collection
    /// mutations on the UI thread.
    public override func _onConstruct() throws {
        _pendingCancellable = _notificationHub.pending
            .sink { [weak self] pending in
                guard let self else { return }
                self.dispatcher.scheduleForeground { [weak self] in
                    self?.syncFromPending(pending)
                }
            }
        try super._onConstruct()
    }

    public override func _onDestruct() throws {
        _pendingCancellable?.cancel()
        _pendingCancellable = nil
        clearVisible()
        try super._onDestruct()
    }

    public override func _onDispose() {
        _pendingCancellable?.cancel()
        _pendingCancellable = nil
        clearVisible()
        super._onDispose()
    }

    // ── Private helpers ────────────────────────────────────────────────────

    /// Mirrors C# `SyncFromPending`: add VMs for new notifications, drop
    /// oldest when over cap, remove VMs whose notifications resolved.
    private func syncFromPending(_ pending: [VMx.Notification]) {
        // Add VMs for new pending notifications, respecting cap.
        for n in pending {
            let key = ObjectIdentifier(n)
            if _map[key] != nil { continue }

            let lifespan = _lifespan ?? 60.0
            let vm = NotificationVM(
                notification: n,
                hub: _notificationHub,
                scheduler: _scheduler,
                lifespan: lifespan
            )
            _map[key] = vm
            _visible.append(vm)

            // Drop oldest while over cap.
            while _visible.count > _cap {
                let oldest = _visible.removeFirst()
                if let oldKey = _map.first(where: { $0.value === oldest })?.key {
                    _map.removeValue(forKey: oldKey)
                }
                oldest.dispose()
            }
        }

        // Remove VMs whose notifications are no longer pending.
        let stillPendingKeys = Set(pending.map { ObjectIdentifier($0) })
        let keysToRemove = _map.keys.filter { !stillPendingKeys.contains($0) }
        for key in keysToRemove {
            if let vm = _map.removeValue(forKey: key) {
                _visible.removeAll { $0 === vm }
                vm.dispose()
            }
        }

        _notifyPropertyChanged("visible")
    }

    private func clearVisible() {
        for vm in _visible { vm.dispose() }
        _visible.removeAll()
        _map.removeAll()
    }

    // ── Builder ────────────────────────────────────────────────────────────

    /// Returns a new empty builder.
    public static func builder() -> NotificationsVMBuilder {
        NotificationsVMBuilder()
    }

    /// Immutable fluent builder for `NotificationsVM`.
    ///
    /// Required: `name(_:)`, `services(hub:dispatcher:)`, `notificationHub(_:)`.
    /// Optional: `hint(_:)`, `scheduler(_:)`, `lifespan(_:)`, `cap(_:)`.
    public struct NotificationsVMBuilder {
        private var _name: String?
        private var _hint: String = ""
        private var _hub: MessageHubProtocol?
        private var _dispatcher: Dispatcher?
        private var _notificationHub: NotificationHubProtocol?
        private var _scheduler: VirtualTimeScheduler?
        private var _lifespan: TimeInterval?
        private var _cap: Int = NotificationsVM.defaultCap

        fileprivate init() {}

        public func name(_ value: String) -> NotificationsVMBuilder {
            var c = self; c._name = value; return c
        }
        public func hint(_ value: String) -> NotificationsVMBuilder {
            var c = self; c._hint = value; return c
        }
        public func services(hub: MessageHubProtocol, dispatcher: Dispatcher) -> NotificationsVMBuilder {
            var c = self; c._hub = hub; c._dispatcher = dispatcher; return c
        }
        public func notificationHub(_ hub: NotificationHubProtocol) -> NotificationsVMBuilder {
            var c = self; c._notificationHub = hub; return c
        }
        public func scheduler(_ scheduler: VirtualTimeScheduler) -> NotificationsVMBuilder {
            var c = self; c._scheduler = scheduler; return c
        }
        /// Overrides the per-notification lifespan (default: 60 s from `NotificationVM`).
        public func lifespan(_ value: TimeInterval) -> NotificationsVMBuilder {
            var c = self; c._lifespan = value; return c
        }
        /// Overrides the cap (default `defaultCap` = 5).
        public func cap(_ value: Int) -> NotificationsVMBuilder {
            var c = self; c._cap = value; return c
        }

        /// Validates required fields and constructs a `NotificationsVM`.
        ///
        /// - Throws: `BuilderValidationError` if any required field is missing.
        public func build() throws -> NotificationsVM {
            guard let name = _name else { throw BuilderValidationError(missingField: "name") }
            guard let hub = _hub else { throw BuilderValidationError(missingField: "hub") }
            guard let dispatcher = _dispatcher else { throw BuilderValidationError(missingField: "dispatcher") }
            guard let notifHub = _notificationHub else { throw BuilderValidationError(missingField: "notificationHub") }
            return NotificationsVM(
                name: name, hint: _hint,
                hub: hub, dispatcher: dispatcher,
                notificationHub: notifHub,
                scheduler: _scheduler ?? VirtualTimeScheduler(),
                lifespan: _lifespan,
                cap: _cap
            )
        }
    }
}
